CRM/Core/BAO/Address.php

```diff
@@ -174,6 +180,22 @@ public static function add(&$params, $fixAddress) {
       //call the function to sync shared address
       self::processSharedAddress($address->id, $params);
 
+      //if address is already shared, share the master address with all children (prevent chaining) CRM-21214
+      if ($address->master_id > 0 && $address->master_id != $address->id) {
+        $result = civicrm_api3('Address', 'get', array(
+          'return' => array("contact_id"),
+          'master_id' => $address->id,
+        ));
+        if ($result['count'] > 0) {
+          $params_copy['address'][1] = $params;
+          foreach ($result['values'] as $key => $value) {
+            $params_copy['contact_id'] = $value['contact_id'];
+            $params_copy['address'][1]['id'] = $key;
+            self::create($params_copy);
+          }
+        }
+      }
+
```

Comment:
There's a small change I'd make here.  If you want to separate this into a separate PR, I'd approve merging the other two changes.

Mainly, I think this should be extracted into its own method.  

I don't even think it necessarily needs a new method - it seems like a second call to `processSharedAddresses()` would do the trick, if that function could optionally not put `master_id` into its `$skipFields` array.  Does that sound right @DanielvV?

Reply:
@MegaphoneJon 

> There's a small change I'd make here. If you want to separate this into a separate PR, I'd approve merging the other two changes.

I'd like to keep it together in one PR.

> Mainly, I think this should be extracted into its own method.

Will do that.

> it seems like a second call to `processSharedAddresses()` would do the trick

I don't think so. This won't call the `processSharedAddressRelationship` method. So for example Person 1 is sharing the address with Person 2 and 3. So using an houshold contact's address for Person 1 creates a relationship between the household and Person 1. It should also create a relationship with person 2 and 3.